kernel/public/test-scenes/0.11.subscribeToEnterLeaveScene1/game.ts

```diff
@@ -0,0 +1,31 @@
+import { engine, Entity, DecentralandInterface, BoxShape, Material, Color3, Transform, Vector3 } from 'decentraland-ecs/src'
+
+//Create entity and assign shape
+const box = new Entity()
+box.addComponent(new BoxShape())
+
+//Create material and configure its fields
+const material = new Material()
+material.albedoColor = Color3.Gray()
+
+//Assign the material to the entity
+box.addComponent(material)
+
+box.addComponent(new Transform({
+  position: new Vector3(8, 0, 8),
+  scale: new Vector3(16, 0.1, 16),
+}))
+
+engine.addEntity(box);
+
+
+declare var dcl: DecentralandInterface
```

Comment:
We try not to expose the `DecentralandInterface` in the SDK, so we should try to create an abstraction similar to the [camera](https://github.com/decentraland/explorer/blob/397acd24d5bdfd764a2d835ec55251332ca40110/kernel/packages/decentraland-ecs/src/decentraland/Camera.ts#L9).

We could create a `Scene` class with some events that could look like this:
```
onEnterScene(listener: ({ userId }) => void ) {
    // subscribe to `onEnterScene` if it hasn't subscribed yet
    dcl.onEvent((evt: any) => {
        if (evt.type === 'onEnterScene') {
            listener(evt.data)
        }
    }) 
}
```
@menduz , does the subscription model make sense like this?
 
And I have one question. If the user starts inside the scene, is the event triggered?

Reply:
I think adding observables in `kernel/packages/decentraland-ecs/src/decentraland/Events.ts` would be a good course of action to start decoupling the `dcl.onEvent` from many functions scattered over there. It also makes the code more testable.

```ts
// kernel/packages/decentraland-ecs/src/decentraland/Events.ts
import { Observable } from '../ecs/Observable'

...
...
...

/**
 * @internal
 * This function generates a callback that is passed to the Observable
 * constructor to subscribe to the events of the DecentralandInterface
 */
function createSubscriber(eventName: keyof IEvents) {
  return () => {
    if (internalDcl) {
      internalDcl.subscribe(eventName)
    }
  }
}

let internalDcl: DecentralandInterface | void

/**
 * @internal
 * This function adds _one_ listener to the onEvent event of dcl interface.
 * Leveraging a switch to route events to the Observable handlers.
 */
export function _initEventObservables(dcl: DecentralandInterface) {
  // store internal reference to dcl, it is going to be used to subscribe to the events
  internalDcl = dcl

  if (internalDcl) {
    internalDcl.onEvent((event) => {
      switch (event.type) {
        case 'onEnterScene': {
          onEnterScene.notifyObservers(event.data)
          return
        }
      }
    })
  }
}



/**
 * These events are triggered after your character enters the scene.
 * @public
 */
export const onEnterScene = new Observable<IEvents['onEnterScene']>(createSubscriber('onEnterScene'))

```

and then call `_initEventObservables(dcl)` from `kernel/packages/decentraland-ecs/src/index.ts`

```diff
/** @internal */
declare let dcl: DecentralandInterface | void
if (typeof dcl !== 'undefined') {
  engine.addSystem(new DecentralandSynchronizationSystem(dcl), Infinity)
+ _initEventObservables(dcl)
}

Reply:
About your question, yes, the event is triggered when you start inside the scene, or when you teleport into the scene.